toolkit/scripts/toolchain.mk

```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
 .SILENT: $(toolchain_rpms)
 
 ifeq ($(REBUILD_TOOLCHAIN),y)
+# We know how to build this archive from scratch
+selected_toolchain_archive = $(final_toolchain)
+else
+# This may be an empty string, that is fine. If its empty we will try to use online packages
+selected_toolchain_archive = $(TOOLCHAIN_ARCHIVE)
+endif
 
+# Extract RPMs from an archive if available, download them from online otherwise.
+ifneq (,$(selected_toolchain_archive))
 # Our manifest files should always track the contents of the freshly built archives exactly
 #   Currently non-blocking, to make this a blocking check switch to `print_error` instead of
 #   `print_warning`
-$(STATUS_FLAGS_DIR)/toolchain_verify.flag: $(TOOLCHAIN_MANIFEST) $(final_toolchain)
-	@echo Validating contents of toolchain against manifest... && \
-	tar -tzf $(final_toolchain) | grep -oP "[^/]+rpm$$" | sort > $(toolchain_actual_contents) && \
+$(STATUS_FLAGS_DIR)/toolchain_verify.flag: $(TOOLCHAIN_MANIFEST) $(selected_toolchain_archive)
+	@echo Validating contents of toolchain against manifest...
+	tar -I $(ARCHIVE_TOOL) -tf $(selected_toolchain_archive) | grep -oP "[^/]+rpm$$" | sort > $(toolchain_actual_contents) && \
 	sort $(TOOLCHAIN_MANIFEST) > $(toolchain_expected_contents) && \
 	diff="$$( comm -3 $(toolchain_actual_contents) $(toolchain_expected_contents) --check-order )" && \
-	echo Toolchain manifest validation complete. && \
 	if [ -n "$${diff}" ]; then \
-		echo "ERROR: Missmatched packages between '$(TOOLCHAIN_MANIFEST)' and '$(final_toolchain)':" && \
+		echo "ERROR: Mismatched packages between '$(TOOLCHAIN_MANIFEST)' and '$(selected_toolchain_archive)':" && \
 		echo "$${diff}"; \
-		$(call print_warning, $@ failed) ; \
-	fi
+		$(call print_error, $@ failed) ; \
```

Comment:
This will detect mismatches in the toolchain manifests (which can have nasty knock-on effects in other build stages). I set this to a warning when I first added it since we were working to stabilize 2.0 at the time. Now that things have calmed down a bit I think we should turn this on.